### DISCUSSION TITLE
.NET Core Linux: Many child processes I didn't start

### DISCUSSION THREAD

**Discussion by dave-yotta** [dotnet/runtime#45835] [Q&A] [answered]
I'm looking at one of our .NET core apps (console - hangfire job server app), running on linux on docker/kubernetes.  I'm just a bit confused as to why we have so many child processes:

```
root@task-executor-85c5557b77-xnrdr:/# pstree 1 -a
sh /start-alloy-engine.sh -x task-executor
  `-dotnet AlloyTaskExecutor.dll
      `-76*[{dotnet}]
```
```
root@task-executor-85c5557b77-xnrdr:/# pstree 1 -a -p -g
sh,1,1 /start-alloy-engine.sh -x task-executor
  `-dotnet,535,1 AlloyTaskExecutor.dll
      |-{dotnet},536,1
      |-{dotnet},537,1
    ....

```
```
htop...
  PID USER      PRI  NI  VIRT   RES   SHR S CPU% MEM%   TIME+  Command
  535 root       20   0 8144M  220M 53612 S 143.  5.6 53:12.41 dotnet AlloyTaskExecutor.dll
  955 root       20   0 8144M  220M 53612 S 37.2  5.6  0:03.81 dotnet AlloyTaskExecutor.dll
  941 root       20   0 8144M  220M 53612 R 37.2  5.6  0:21.58 dotnet AlloyTaskExecutor.dll
  614 root       20   0 8144M  220M 53612 S 27.2  5.6 10:32.30 dotnet AlloyTaskExecutor.dll
  599 root       20   0 8144M  220M 53612 S 17.9  5.6  6:53.90 dotnet AlloyTaskExecutor.dll
  655 root       20   0 8144M  220M 53612 S 12.0  5.6  3:02.22 dotnet AlloyTaskExecutor.dll
  649 root       20   0 8144M  220M 53612 S  6.6  5.6  3:05.03 dotnet AlloyTaskExecutor.dll
  660 root       20   0 8144M  220M 53612 S  5.3  5.6  3:02.42 dotnet AlloyTaskExecutor.dll
  946 root       20   0 8144M  220M 53612 S  0.0  5.6  0:13.86 dotnet AlloyTaskExecutor.dll
  960 root       20   0 8144M  220M 53612 S  0.0  5.6  0:02.78 dotnet AlloyTaskExecutor.dll
  541 root       20   0 8144M  220M 53612 S  0.0  5.6  0:00.47 dotnet AlloyTaskExecutor.dll
    1 root       20   0  2388   408   296 S  0.0  0.0  0:00.11 /bin/sh /start-alloy-engine.sh -x task-executor
  536 root       20   0 8144M  220M 53612 S  0.0  5.6  0:00.00 dotnet AlloyTaskExecutor.dll
  537 root       20   0 8144M  220M 53612 S  0.0  5.6  0:00.00 dotnet AlloyTaskExecutor.dll
....
```

There are a lot of `dotnet AlloyTaskExecutor.dll` processes that are all children of the one I created.  This is happening on all our apps when run on linux, not seeing it on windows.

I'm looking into resource usage of this app here, so it's a bit confusing.

### Is this some kind of way multitasking is done on .net core linux? 

I may want to do some profiling - which process do I even attach to? Just PID 535?

### Is it something we've done?

We have no Process.Start in our codebase, so it may be a 3rd party lib doing this.

They are all trying to run our app with default args?  Is there a different way to change the entrypoint I'm not seeing here? 

Any advice to find the culprit? I've thought of changing the process security to deny process spawning but I don't know how to do this on linux.

`ps aux` and `top` are only showing one process, not sure why that is either.

Are these forked processes maybe?  If it helps the app is running hangfire with a redis storage provider.

**Comment by PathogenDavid** [comment, marked as answer]
These are just individual threads within the same process. In htop if you press F2 you can hide them, color them differently, and show their custom thread names. (IE: [`Thread.Name`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.name?view=net-5.0#System_Threading_Thread_Name))

For example, here's what it looks like if you enable coloring them differently and showing thread names:

![image](https://user-images.githubusercontent.com/278957/101639748-460d9c80-39f5-11eb-97c7-cb8e1613a923.png)

The reason things look the way they do by default is the line between a thread and a process is slightly blurrier on Linux. ([This elaborates a bit more](https://stackoverflow.com/a/8787888/636752))
